contracts/utils/MultiRateLimited.sol

```diff
@@ -0,0 +1,170 @@
+import "@openzeppelin/contracts/utils/math/Math.sol";
+
+/// @title contract for putting a rate limit on how fast an address can perform an action e.g. Minting
+/// there are two buffers, one buffer which is each individual addresses's current buffer,
+/// and then there is a global buffer which is the buffer that each individual address must respect as well
+/// @author Fei Protocol
+contract MultiRateLimited is RateLimited, IMultiRateLimited {
+
+    /// @notice the rate per second for each address
+    mapping (address => uint256) public individualRateLimitPerSecond;
+
+    /// @notice the last time the buffer was used by each address 
+    mapping (address => uint256) public individualLastBufferUsedTime;
+
+    /// @notice the cap of the buffer that can be used at once by a single address
+    mapping (address => uint256) public individualBufferCap;
+
+    /// @notice the buffer at the timestamp of lastBufferUsedTime
+    mapping (address => uint256) private _bufferStored;
+
+    constructor(
+        address coreAddress,
+        uint256 _maxRateLimitPerSecond,
+        uint256 _rateLimitPerSecond,
+        uint256 _bufferCap,
+        uint256 maxBufferCap,
+        bool _doPartialAction
+    )
+        CoreRef(coreAddress)
+        RateLimited(_maxRateLimitPerSecond, _rateLimitPerSecond, _bufferCap, _doPartialAction)
+    {}
+
+    /// @notice add an authorized minter contract
+    /// @param _minter the new address to add as a minter
+    /// @param _rateLimitPerSecond the rate limit per second for this minter
+    /// @param _bufferCap  the buffer cap for this minter
+    function addAddress(address _minter, uint256 _rateLimitPerSecond, uint256 _bufferCap) external virtual override onlyGovernorOrAdmin {
+        require(_bufferCap <= bufferCap, "MultiRateLimited: new buffercap too high");
+
+        individualRateLimitPerSecond[_minter] = _rateLimitPerSecond;
+        individualLastBufferUsedTime[_minter] = block.timestamp;
+        individualBufferCap[_minter] = _bufferCap;
+        _bufferStored[_minter] = _bufferCap;
+
+        emit IndividualRateLimitPerSecondUpdate(_minter, 0, _rateLimitPerSecond);
+    }
+
+    /// @notice add an authorized minter contract
+    /// @param _minter the address whose buffer and rate limit per second will be set
+    /// @param _rateLimitPerSecond the new rate limit per second for this minter
+    /// @param _bufferCap  the new buffer cap for this minter
+    function updateAddress(address _minter, uint256 _rateLimitPerSecond, uint256 _bufferCap) external virtual override onlyGovernorOrAdmin {
+        require(_bufferCap <= bufferCap, "MultiRateLimited: buffercap too high");
+
+        individualRateLimitPerSecond[_minter] = _rateLimitPerSecond;
+        individualLastBufferUsedTime[_minter] = block.timestamp;
+        individualBufferCap[_minter] = _bufferCap;
+        _bufferStored[_minter] = _bufferCap;
+
+        emit IndividualRateLimitPerSecondUpdate(_minter, 0, _rateLimitPerSecond);
```

Comment:
On this event, the second argument shouldn't be `0` right? It is `uint256 oldRateLimitPerSecond` so should be `individualRateLimitPerSecond[_minter]`?

Reply:
fixed